src/lotus/client.rs

```diff
@@ -0,0 +1,154 @@
+use crate::types::{
+    CIDMap, MpoolPushMessage, MpoolPushMessageInner, MpoolPushMessageResponse, ReadStateResponse,
+    StateWaitMsgResponse, WalletKeyType, WalletListResponse,
+};
+use crate::JsonRpcClient;
+use crate::LotusApi;
+use anyhow::Result;
+use async_trait::async_trait;
+use cid::Cid;
+use fvm_shared::address::Address;
+use fvm_shared::econ::TokenAmount;
+use num_traits::cast::ToPrimitive;
+use serde::de::DeserializeOwned;
+use serde_json::json;
+use std::fmt::Debug;
+use std::str::FromStr;
+
+// RPC endpoints
+mod endpoints {
+    pub const MEM_PUSH_MESSAGE_ENDPOINT: &str = "Filecoin.MpoolPushMessage";
+    pub const STATE_WAIT_MSG: &str = "Filecoin.StateWaitMsg";
+    pub const WALLET_NEW: &str = "Filecoin.WalletNew";
+    pub const WALLET_LIST: &str = "Filecoin.WalletList";
+    pub const WALLET_DEFAULT_ADDRESS: &str = "Filecoin.WalletDefaultAddress";
+    pub const READ_STATE: &str = "Filecoin.StateReadState";
+}
+
+/// The lotus client
+pub struct LotusClient<Inner> {
+    inner: Inner,
+}
+
+impl<Inner> LotusClient<Inner> {
+    pub fn new(inner: Inner) -> Self {
+        Self { inner }
+    }
+}
+
+#[async_trait]
+impl<Inner: JsonRpcClient + Send + Sync> LotusApi for LotusClient<Inner> {
+    async fn mpool_push_message(&self, msg: MpoolPushMessage) -> Result<MpoolPushMessageInner> {
+        let from = if let Some(f) = msg.from {
+            f
+        } else {
+            self.wallet_default().await?
+        };
+
+        let nonce = msg
+            .nonce
+            .map(|n| serde_json::Value::Number(n.into()))
+            .unwrap_or(serde_json::Value::Null);
+
+        let f = |t: Option<TokenAmount>| {
+            t.map(|n| serde_json::Value::Number(n.atto().to_u64().unwrap().into()))
+                .unwrap_or(serde_json::Value::Null)
+        };
+        let gas_limit = f(msg.gas_limit);
+        let gas_premium = f(msg.gas_premium);
+        let gas_fee_cap = f(msg.gas_fee_cap);
+        let max_fee = f(msg.max_fee);
+
+        // refer to: https://lotus.filecoin.io/reference/lotus/mpool/#mpoolpushmessage
+        let to_send = json!([
```

Comment:
These are the `params` to the JSON_RPC request. Please rename appropriately, here and elsewhere through this file.